FFmpegInterop/Metadata.cpp

```diff
@@ -18,39 +18,140 @@
 
-		// TODO: Populate the property handler with the metadata
+		return propHandler;
 	}
 
-	void SetMSSThumbnail(_In_ const MediaStreamSource& mss, _In_ const AVStream* thumbnailStream)
+	void PopulateMetadata(_In_ const MediaStreamSource& mss, _In_opt_ const AVDictionary* metadata)
 	{
-		// Write the thumbnail to a stream
-		AVPacket_ptr packet{ av_packet_clone(&thumbnailStream->attached_pic) };
+		if (metadata == nullptr)
+		{
+			return;
+		}
+
+		com_ptr<IPropertyStore> propHandler{ GetPropertyHandler(mss) };
+#if DBG
+		com_ptr<IPropertyStoreCapabilities> propHandlerCapabilties{ propHandler.as<IPropertyStoreCapabilities>() };
+#endif // DBG
+
+		const AVDictionaryEntry* tag{ nullptr };
+		while(tag = av_dict_iterate(metadata, tag))
+		{
+			FFMPEG_INTEROP_TRACE("Property: Key = %hs, Value = %hs", tag->key, tag->value);
+
+			// Check if we have a mapping for the property
+			auto iter{ c_metadataPropertiesMap.find(tolower(tag->key)) };
+			if (iter != c_metadataPropertiesMap.end())
+			{
+				const PropertyKey& propKey{ iter->second };
+				unique_prop_variant prop;
+				prop.vt = propKey.type;
+
+				// Convert/parse the property value
+				try
+				{
+					switch (propKey.type)
+					{
+					case VT_LPWSTR:
+						prop.pwszVal = to_cotaskmem_string(tag->value).release();
+						break;
+
+					case VT_UI4:
+						prop.ulVal = stoul(tag->value);
+						break;
+
+					default:
+						WINRT_ASSERT(false);
+						continue;
+					}
+
+#if DBG
+					// The MSS property handler marks some properties read-only which prevents us from setting them
+					WINRT_ASSERT(propHandlerCapabilties->IsPropertyWritable(propKey.key) == S_OK); // S_FALSE indicates the property is read-only
+#endif // DBG
+					THROW_IF_FAILED(propHandler->SetValue(propKey.key, prop));
+				}
+				catch (...)
+				{
+					LOG_CAUGHT_EXCEPTION_MSG("Failed to convert/parse/set the property value");
+					continue;
+				}
+
+			}
+		}
+	}
+
+	void SetThumbnail(
+		_In_ const MediaStreamSource& mss,
+		_In_ const AVStream* stream,
+		_In_opt_ const FFmpegInterop::FFmpegInteropMSSConfig& config)
+	{
+		WINRT_ASSERT(stream->disposition == AV_DISPOSITION_ATTACHED_PIC);
+
+		// Don't set a thumbnail on the MSS in media source app service scenarios. The remote stream reference could
+		// outlive the app service connection. If the remote stream reference is released in the client app after the
+		// app service has been suspended, then the remote release will block until the app service is terminated. This
+		// could potentially cause the client app to become unresponsive during this time. 
+		if (config != nullptr && config.IsMediaSourceAppService())
+		{
+			FFMPEG_INTEROP_TRACE("Stream %d: Not setting thumbnail because this is a media source app service", stream->index);
+			return;
+		}
+
+		FFMPEG_INTEROP_TRACE("Stream %d: Setting thumbnail", stream->index);
+
+		stream->attached_pic.data;

```

Comment:
Why is stream->attached_pic.data called here, twice?

Reply:
why even once? :)

Reply:
Unintentional. I must have stopped in the middle of writing something and forgot to clean these lines up.